File: MyMonitorHub.JavaAgent/src/main/java/mymonitorhub/agent/core/AbstractPlugin.java
package mymonitorhub.agent.core;

import com.fasterxml.jackson.databind.JsonNode;
import mymonitorhub.agent.model.EventType;
import mymonitorhub.agent.model.ItemStyle;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Date;

/**
 * Base class for all monitoring plugins.
 * Mirrors C# AbstractPlugin.
 */
public abstract class AbstractPlugin {

    private static final Logger log = LoggerFactory.getLogger(AbstractPlugin.class);

    private int hibernateSeconds;
    private Date currentRunTime = new Date(0);
    private JsonNode pluginElement;
    private String title;

    public String getTitle() { return title; }

    public int getHibernateSeconds() { return hibernateSeconds; }

    public void setHibernateSeconds(int value) {
        hibernateSeconds = value;
        JsonNode hibernateNode = pluginElement != null ? pluginElement.get("hibernate-seconds") : null;
        if (hibernateNode != null && !hibernateNode.isNull()) {
            int fromConfig = hibernateNode.asInt(value);
            hibernateSeconds = Math.max(fromConfig, value);
        }
    }

    protected JsonNode getPluginElement() { return pluginElement; }

    public void initialize(JsonNode element, String threadTitle) {
        this.pluginElement = element;
        String pluginTitle = element != null ? element.path("title").asText(null) : null;
        if (pluginTitle != null) {
            this.title = pluginTitle.equals(threadTitle) ? pluginTitle : threadTitle + ">" + pluginTitle;
        } else {
            this.title = threadTitle;
        }
        setHibernateSeconds(120);
    }

    public boolean canRun() {
        if (currentRunTime.getTime() == 0) return true;
        long millis = new Date().getTime() - currentRunTime.getTime();
        long seconds = millis / 1000;
        return seconds >= hibernateSeconds;
    }

    public void start() {
        currentRunTime = new Date();
        log.debug("starting");
        execute();
    }

    protected abstract void execute();

    protected boolean sayStatus(String category, String subCategory, String name,
                                String description, EventType status) {
        return sayStatus(category, subCategory, name, description, status, ItemStyle.SendsHealth);
    }

    protected boolean sayStatus(String category, String subCategory, String name,
                                String description, EventType status, ItemStyle style) {
        return Monitor.fireEvent(category, subCategory, name, description, status, style, currentRunTime);
    }

    /**
     * Always fires the event unconditionally, bypassing the status-change
     * deduplication used by {@link #sayStatus}. Use when a state transition
     * must be reported regardless of the previous value (e.g. a counter reset).
     */
    protected final void sayStatusAlways(String category, String subCategory, String name,
                                         String description, EventType status) {
        sayStatusAlways(category, subCategory, name, description, status, ItemStyle.SendsHealth);
    }

    protected final void sayStatusAlways(String category, String subCategory, String name,
                                         String description, EventType status, ItemStyle style) {
        Monitor.fireEventAlways(category, subCategory, name, description, status, style);
    }
}

File: MyMonitorHub.Agent/MyMonitorHub.Agent.Core/Core/AbstractPlugin.cs
using System;
using MyMonitorHub.Common.WebApi;
using Newtonsoft.Json.Linq;
using Serilog;

namespace MyMonitorHub.Agent.Core
{
    public abstract class AbstractPlugin
    {
        private static readonly ILogger Logger = Log.ForContext<AbstractPlugin>();
        private int _hibernateSeconds;
        private DateTime _currentRunTime;

        private JToken? _pluginElement;
        private string? _title;

        public string? Title => _title;

        public int HibernateSeconds
        {
            set
            {
                _hibernateSeconds = value;

                var hibernateValue = _pluginElement?["hibernate-seconds"];
                if (hibernateValue != null)
                {
                    _hibernateSeconds = hibernateValue.Value<int>();
                    if (_hibernateSeconds < value)
                    {
                        _hibernateSeconds = value;
                    }
                }
            }
            get { return _hibernateSeconds; }
        }

        protected JToken? PluginElement => _pluginElement;

        public virtual void Initialize(JToken pluginElement, string threadTitle)
        {
            _pluginElement = pluginElement;
            var pluginTitle = pluginElement?["title"]?.ToString();

            if (pluginTitle != null)
            {
                if (pluginTitle == threadTitle)
                {
                    _title = pluginTitle;
                }
                else
                {
                    _title = threadTitle + ">" + pluginTitle;
                }
            }

            HibernateSeconds = 120; // default is 2 minutes minimum. Can be set larger in monitor.json

        }

        public bool CanRun()
        {
            if (_currentRunTime == DateTime.MinValue)
            {
                return true;
            }
            var now = DateTime.Now;
            var millis = (now - _currentRunTime).TotalMilliseconds;
            var seconds = (int) (millis/1000);
            if (seconds >= _hibernateSeconds)
            {
                return true;
            }
            return false;
        }

        public void Start()
        {
            _currentRunTime = DateTime.Now;
            Logger.Debug("starting");
            Execute();
        }

        protected abstract void Execute();


        protected bool SayStatus(string category, string subCategory, string name, string description,
                                          EventType status, ItemStyle health = ItemStyle.SendsHealth)
        {
            return Monitor.FireEvent(category, subCategory, name, description, status, health,_currentRunTime);
        }
    }
}